### PR TITLE
planner: Support assign `DEFAULT` in `ON DUPLICATE KEY UPDATE` statement

### LINKED ISSUE #13071: Support assign `DEFAULT` in `ON DUPLICATE KEY UPDATE` statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t1 (a int unique, 
	b int generated always as (-a) virtual,
	c int generated always as (-a) stored);


insert into t1 values (1,default,default);

insert into t1 values (1,default,default) on duplicate key update a=2, b=default;
```
2. What did you expect to see?

```
Query OK, 1 row affected (0.01 sec)
```

3. What did you see instead?

```
ERROR 3105 (HY000): The value specified for generated column 'b' in table 't1' is not allowed.
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
Release Version: v4.0.0-alpha-693-gada724a2a
```

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 
Fix #13071 

### What is changed and how it works?

In statement `insert into t value (1) on duplicate key update a=default`, when we build the `InsertPlan`, we also check:

If the column need to be updated on duplicate key is generated column?
- True: if the assigned variable is `default`?
	- True: we ignore this assignment.
	- False: throw the error `The value specified for generated column 'gc' in table 't' is not allowed.`
- False: add the normal assignment to `insertPlan.OnDuplicate`

### Check List 

Tests 

 - Integration test

Code changes

 - None

Side effects

 - Increased code complexity

Related changes

 - This PR is part of #11901 

Release note

 - Support assign `DEFAULT` in `ON DUPLICATE KEY UPDATE` statement

### CODE REVIEW

**Comment by Deardrops** [comment]
> Please check `insert into t1 values (1,default,default) on duplicate key update a=2, b=default(t);`

@wjhuang2016 Comment addressed, PTAL

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3101,6 +3101,17 @@ func (b *PlanBuilder) buildUpdateLists(
 	return newList, p, allAssignmentsAreConstant, nil
 }
 
+// extractDefaultExpr extract a `DefaultExpr` without any parameter from a `ExprNode`
+// return it and if successful extract it
+// Note the sql function `DEFAULT(a)` is not same with keyword `DEFAULT`
+// Sql function `DEFAULT(a)` will return `false`
+func extractDefaultExpr(node ast.ExprNode) (*ast.DefaultExpr, bool) {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: Can we only return one value, which is the `expr`? the second bool value can be derived by `expr == nil`

Comment (Deardrops) [reply planner/core/logical_plan_builder.go]: I think this way is easier to understand:
```go
if expr, ok := extractDefaultExpr(assign.Expr); ok {
	expr.Name = assign.Column
}
```
vs
```go
if expr := extractDefaultExpr(assign.Expr); expr != nil {
	expr.Name = assign.Column
}
```

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -1941,14 +1949,15 @@ func (b *PlanBuilder) buildSetValuesOfInsert(ctx context.Context, insert *ast.In
...
+		isDefault := false
+		if expr := extractDefaultExpr(assign.Expr); expr != nil {
+			expr.Name = assign.Column
+			isDefault = true
 		}
 		// Note: For INSERT, REPLACE, and UPDATE, if a generated column is inserted into, replaced, or updated explicitly, the only permitted value is DEFAULT.
 		// see https://dev.mysql.com/doc/refman/8.0/en/create-table-generated-columns.html
 		if _, ok := generatedColumns[assign.Column.Name.L]; ok {
-			if isDefaultExpr {
+			if isDefault {
```
Comment (wjhuang2016) [inline planner/core/planbuilder.go]: How about:
if defaultExpr := extractDefaultExpr(assign.Expr); defaultExpr != nil {
			defaultExpr.Name = assign.Column				expr.Name = assign.Column
		}			}
...
if defaultExpr != nil {
}

Comment (Deardrops) [reply planner/core/planbuilder.go]: comment addressed, PTAL @wjhuang2016

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -1866,15 +1866,14 @@ func (p *Insert) resolveOnDuplicate(onDup []*ast.Assignment, tblInfo *model.Tabl
 
 		// Check whether the column to be updated is the generated column.
 		column := colMap[assign.Column.Name.L]
-		isDefault := false
-		if expr := extractDefaultExpr(assign.Expr); expr != nil {
-			expr.Name = assign.Column
-			isDefault = true
+		var defaultExpr *ast.DefaultExpr
+		if defaultExpr = extractDefaultExpr(assign.Expr); defaultExpr != nil {
+			defaultExpr.Name = assign.Column
```
Comment (wjhuang2016) [inline planner/core/planbuilder.go]: defaultExpr := extractDefaultExpr(assign.Expr); 
if defaultExpr != nil {
...
}

Comment (Deardrops) [reply planner/core/planbuilder.go]: @wjhuang2016 Address comment, PTAL

**Pull request merged**